Send buffer visit / unload events to completer server

AndyMoreland opened the pull request:
The typescript completion service requires BufferVisit and BufferUnload events to provide semantic results, so I've attempted to implement support for them in this patch.

The general goal is just to send a `BufferVisit` event when a buffer is first opened, and to send a `BufferUnload` event when a buffer is closed. 

This patch seems to work for me, but I am not very familiar at all with `ycmd` or the intricacies of minor mode initialization ordering, so it might have race conditions. I'm happy to make any changes required.

ptrv commented:
Hi Andy!

Thanks fro your PR! I have some comments on the changes.

ptrv commented on ycmd.el:
```diff
@@ -551,7 +551,8 @@ and `delete-process'.")
     (kill-buffer-hook                 . ycmd--teardown)
-    (before-revert-hook               . ycmd--teardown))
+    (before-revert-hook               . ycmd--teardown)
+    (hack-local-variables-hook        . ycmd--on-open))
```
Why `hack-local-variables-hook`? Wouldn't it be enough to cann `ycmd--on-open` when the mode is activated? Basically calling the function in the body of the minor mode definition? Or in the minor mode `after-hook` before calling `(ycmd--comditional-parse 'mode-enabled)`?

AndyMoreland replied:
After I got in bed I started thinking the same thing :-). I was miming from `tide-mode`, which is the typescript-ide mode, and I think this method made more sense in that context. I'll make the change.

ptrv commented on ycmd.el:
```diff
@@ -1462,6 +1467,40 @@ Otherwise the response is probably an exception."
     (ycmd--report-status 'parsed)
     (run-hook-with-args 'ycmd-file-parse-result-hook results)))
 
+(defun ycmd-notify-buffer-visit ()
```
These two functions are almost the same. You could make a function i.e. `ycmd--notify-server` which takes one argument for the `event_name` and wrap the function once with `BufferVisit` and once with `BufferUnload`. Also the two functions should have the prefix `ycmd--` because the should be marked as `private`.u

abingham commented:
This looks good. If you make the changes that @ptrv mentioned, then I think it'll be ready to merge in. Thanks!

AndyMoreland commented:
I had to rip the `notify` code out of `ycmd--teardown` because it was causing infinite recursion when I invoked `ycmd-mode` interactively. It seems `ycmd-mode` executes `ycmd--teardown`, and when that triggers a `ycmd--request`, the stack gets blown. Otherwise I believe I've addressed the comments in my second commit.

ptrv commented:
@AndyMoreland Nice. Thanks! I have still some minor comments :smile:

ptrv commented on ycmd.el:
```diff
@@ -649,14 +649,16 @@ Otherwise behave as if called interactively.
-      (remove-hook (car hook) (cdr hook) 'local))
-    (ycmd--teardown))))
+  (progn
+    (ycmd--notify-server "BufferVisit")
```
I think it's better to put this call in the `:after-hook` above:

```emacs-lisp
(progn
  (ycmd--notify-server "BufferVisit")
  (ycmd--conditional-parse 'mode-enabled))
```

ptrv commented on ycmd.el:
```diff
@@ -740,6 +742,10 @@ _LEN is ununsed."
        (when ycmd-mode
          ,@body))))
 
+(defun ycmd--close-buffer ()
```
Please add a small comment

ptrv commented on ycmd.el:
```diff
@@ -1462,6 +1468,22 @@ Otherwise the response is probably an exception."
     (run-hook-with-args 'ycmd-file-parse-result-hook results)))
 
+(defun ycmd--notify-server (event-name)
+  "Send a simple event notification to the server, ignoring response."
```
`event-name` needs to be mentioned in the docstring with capital letters i.e. `EVENT-NAME`

ptrv commented on ycmd.el:
```diff
@@ -1462,6 +1468,22 @@ Otherwise the response is probably an exception."
+                     :parser 'json-read)
+      (deferred:error it
+        (lambda (err)
+          (message "Error sending buffer visit request: %s" err)
```
The comment is not generic, please change to:

```emacs-lisp
(message "Error sending %s request: %s" event-name err)
```

ptrv commented:
After applying the changes please also squash the commits into one commit. Then we can merge.

ptrv commented:
hmm, I am getting a python `KeyError: u'unloaded_buffer` when closing a `cpp` file with this patch:

```
2016-07-12 13:41:33,653 - DEBUG - Event name: BufferUnload
Traceback (most recent call last):
  File "/Users/peter.vasil/src/ycmd/third_party/bottle/bottle.py", line 861, in _handle
    return route.call(**args)
  File "/Users/peter.vasil/src/ycmd/third_party/bottle/bottle.py", line 1734, in wrapper
    rv = callback(*a, **ka)
  File "/Users/peter.vasil/src/ycmd/ycmd/../ycmd/watchdog_plugin.py", line 106, in wrapper
    return callback( *args, **kwargs )
  File "/Users/peter.vasil/src/ycmd/ycmd/../ycmd/hmac_plugin.py", line 72, in wrapper
    body = callback( *args, **kwargs )
  File "/Users/peter.vasil/src/ycmd/ycmd/../ycmd/handlers.py", line 65, in EventNotification
    event_handler )( request_data )
  File "/Users/peter.vasil/src/ycmd/ycmd/../ycmd/completers/cpp/clang_completer.py", line 337, in OnBufferUnload
    ToCppStringCompatible( request_data[ 'unloaded_buffer' ] ) )
  File "/Users/peter.vasil/src/ycmd/ycmd/../ycmd/request_wrap.py", line 83, in __getitem__
    return self._request[ key ]
KeyError: u'unloaded_buffer'
2016-07-12 13:41:53,396 - INFO - Received health request
```

ptrv commented:
I guess we need to pass also the unlaoded buffer path(?) to the server when calling `BufferUnload`

ptrv commented:
Ok, I filed an issue in `ycmd` project to just use the `filepath` that is already specified in the request data instead of having to specify `unloaded_buffer`.

See issue: https://github.com/Valloric/ycmd/issues/541

I think we should wait for them until this can get merged.

AndyMoreland commented:
re: the resolution in the `ycmd` issue: I'll update this PR to do the right thing.

ptrv commented:
@AndyMoreland Thanks for applying the suggestions. For the buffer unload issue we will need to pass also `unloaded_buffer` alongside with the standard content.

AndyMoreland commented:
yep! working on it.

AndyMoreland commented:
I've added a new commit to pass `unloaded_buffer` with the standard content. If that looks good I'm happy to squash it in.

ptrv commented:
@AndyMoreland Looks good! Please squash the two commits into one and the we're happy to merge it.

AndyMoreland commented:
Done.

ptrv commented:
Merged via 36309931f2884e85c0635522f061cf149fa39665

Thanks for the PR!

I had to update the commit a bit because one test was failing.

 `BufferVisit` is called now before `FileReadyToParse` when the new flag `ycmd--buffer-visit-flag` is `nil`. We need to handle also the case when we restart the server but not closing the buffer. Then we need to send `BufferVisit` again after the server started before parsing the buffer.

Also we should only send a `BufferUnload` notification when the server is running. Otherwise `ycmd--request` would restart the server which doesn't make sense at that point.

ptrv commented:
This feature is also really nice for the `clang_completer` because a `BufferUnload` notification deletes the caches for that file.

AndyMoreland commented:
Thank you :-)
